earthpy/tests/test_clip.py

```diff
@@ -19,33 +19,47 @@ def make_locs_gdf():
     return gdf
 
 
-def make_poly_in_gdf():
+def make_single_rect_poly_gdf():
     """ Bounding box polygon. """
     poly_inters = Polygon([(0, 0), (0, 10), (10, 10), (10, 0), (0, 0)])
     gdf = gpd.GeoDataFrame(
         [1], geometry=[poly_inters], crs={"init": "epsg:4326"}
     )
+    gdf["attr2"] = ["site-boundary"]
+    return gdf
+
+
+def make_smaller_clip_rect_poly_gdf():
+    """ Bounding box polygon. """
+    poly_inters = Polygon([(-5, -5), (-5, 15), (15, 15), (15, -5), (-5, -5)])
+    gdf = gpd.GeoDataFrame(
+        [1], geometry=[poly_inters], crs={"init": "epsg:4326"}
+    )
+    gdf["attr2"] = ["study area"]
     return gdf
 
 
 def make_locs_buff():
     """ Buffer points to create multi poly. """
     buffered_locations = make_locs_gdf()
     buffered_locations["geometry"] = buffered_locations.buffer(4)
+    buffered_locations["type"] = "plot"
     return buffered_locations
 
 
 def make_donut_geom():
     """ Make a donut geometry. """
     donut = gpd.overlay(
-        make_locs_buff(), make_poly_in_gdf(), how="symmetric_difference"
+        make_locs_buff(),
+        make_single_rect_poly_gdf(),
+        how="symmetric_difference",
     )
     return donut
 
 
 @pytest.fixture
 def locs_gdf():
-    """ Get a dummy point GeoDataFrame.
+    """ Create a sample point GeoDataFrame.
 
     This fixture calls make_locs_gdf(), which is a function that is used in
```

Comment:
@mbjoseph  this is where i have the question. i'm really not sure how to translate this in terms of cleaning up the fixtures here... can you somehow clarify for me? that discussion actually confused me more than it helped.

Reply:
Sorry that discussion was not helpful. If you're looking to get rid of the helper functions, you could define some fixtures in terms of others, which would definitely simplify things, e.g., 

```python
"""Tests for the clip module."""

import pytest
import numpy as np
from shapely.geometry import Polygon, Point, LineString
import shapely
import geopandas as gpd
import earthpy.clip as cl


@pytest.fixture
def locs_gdf():
    """ Create a sample point GeoDataFrame.

    This fixture calls make_locs_gdf(), which is a function that is used in
    multiple fixtures. But, fixtures are not supposed to be used like that:

    see https://github.com/pytest-dev/pytest/issues/3950 for discussion
    """
    pts = np.array([[2, 2], [3, 4], [9, 8], [-12, -15]])
    gdf = gpd.GeoDataFrame(
        [Point(xy) for xy in pts],
        columns=["geometry"],
        crs={"init": "epsg:4326"},
    )
    return gdf


@pytest.fixture
def locs_buff(locs_gdf):
    """ Buffer points to create multi poly. """
    buffered_locations = locs_gdf
    buffered_locations["geometry"] = buffered_locations.buffer(4)
    buffered_locations["type"] = "plot"
    return buffered_locations


@pytest.fixture
def linez_gdf():
    """ Create Line Objects For Testing """
    linea = LineString([(1, 1), (2, 2), (3, 2), (5, 3)])
    lineb = LineString([(3, 4), (5, 7), (12, 2), (10, 5), (9, 7.5)])
    gdf = gpd.GeoDataFrame(
        [1, 2], geometry=[linea, lineb], crs={"init": "epsg:4326"}
    )
    return gdf


@pytest.fixture
def single_rect_poly_gdf():
    """ Fixture for a bounding box polygon. """
    poly_inters = Polygon([(0, 0), (0, 10), (10, 10), (10, 0), (0, 0)])
    gdf = gpd.GeoDataFrame(
        [1], geometry=[poly_inters], crs={"init": "epsg:4326"}
    )
    gdf["attr2"] = ["site-boundary"]
    return gdf


@pytest.fixture
def smaller_clip_rect_poly_gdf():
    """ Bounding box from (-5, 5) to (15, 15). """
    poly_inters = Polygon([(-5, -5), (-5, 15), (15, 15), (15, -5), (-5, -5)])
    gdf = gpd.GeoDataFrame(
        [1], geometry=[poly_inters], crs={"init": "epsg:4326"}
    )
    gdf["attr2"] = ["study area"]
    return gdf


@pytest.fixture
def donut_geom(locs_buff, single_rect_poly_gdf):
    """ Fixture for donut geometry objects. """
    donut = gpd.overlay(
        locs_buff,
        single_rect_poly_gdf,
        how="symmetric_difference",
    )
    return donut


@pytest.fixture
def multi_gdf(donut_geom):
    """ Create a multi-polygon GeoDataFrame. """
    multi_poly = donut_geom.unary_union
    out_df = gpd.GeoDataFrame(
        geometry=gpd.GeoSeries(multi_poly), crs={"init": "epsg:4326"}
    )
    out_df = out_df.rename(columns={0: "geometry"}).set_geometry("geometry")
    out_df["attr"] = ["pool"]
    return out_df


@pytest.fixture
def multi_line(linez_gdf):
    """ Create a multi-line GeoDataFrame.

    This has one multi line and another regular line.
    """
    # Create a single and multi line object
    multiline_feat = linez_gdf.unary_union
    linec = LineString([(2, 1), (3, 1), (4, 1), (5, 2)])
    out_df = gpd.GeoDataFrame(
        geometry=gpd.GeoSeries([multiline_feat, linec]),
        crs={"init": "epsg:4326"},
    )
    out_df = out_df.rename(columns={0: "geometry"}).set_geometry("geometry")
    out_df["attr"] = ["road", "stream"]
    return out_df


@pytest.fixture
def multi_point(locs_gdf):
    """ Create a multi-line GeoDataFrame. """
    multi_point = locs_gdf.unary_union
    out_df = gpd.GeoDataFrame(
        gpd.GeoSeries(
            [multi_point, Point(2, 5), Point(-11, -14), Point(-10, -12)]
        ),
        crs={"init": "epsg:4326"},
    )
    out_df = out_df.rename(columns={0: "geometry"}).set_geometry("geometry")
    out_df["attr"] = ["tree", "another tree", "shrub", "berries"]
    return out_df


def test_not_gdf(single_rect_poly_gdf):
    """Non-GeoDataFrame inputs raise attribute errors."""
    with pytest.raises(AttributeError):
        cl.clip_shp((2, 3), single_rect_poly_gdf)
    with pytest.raises(AttributeError):
        cl.clip_shp(single_rect_poly_gdf, (2, 3))


def test_returns_gdf(locs_gdf, single_rect_poly_gdf):
    """Test that function returns a GeoDataFrame (or GDF-like) object."""
    out = cl.clip_shp(locs_gdf, single_rect_poly_gdf)
    assert hasattr(out, "geometry")


def test_non_overlapping_geoms():
    """Test that a bounding box returns error if the extents don't overlap"""
    unit_box = Polygon([(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)])
    unit_gdf = gpd.GeoDataFrame(
        [1], geometry=[unit_box], crs={"init": "epsg:4326"}
    )
    non_overlapping_gdf = unit_gdf.copy()
    non_overlapping_gdf = non_overlapping_gdf.geometry.apply(
        lambda x: shapely.affinity.translate(x, xoff=20)
    )
    with pytest.raises(ValueError):
        cl.clip_shp(unit_gdf, non_overlapping_gdf)


def test_input_gdfs(single_rect_poly_gdf):
    """Test that function fails if not provided with 2 GDFs."""
    with pytest.raises(AttributeError):
        cl.clip_shp(list(), single_rect_poly_gdf)
    with pytest.raises(AttributeError):
        cl.clip_shp(single_rect_poly_gdf, list())


def test_clip_points(locs_gdf, single_rect_poly_gdf):
    """Test clipping a points GDF with a generic polygon geometry."""
    clip_pts = cl.clip_shp(locs_gdf, single_rect_poly_gdf)
    assert len(clip_pts.geometry) == 3 and clip_pts.geom_type[1] == "Point"


def test_clip_poly(locs_buff, single_rect_poly_gdf):
    """Test clipping a polygon GDF with a generic polygon geometry."""
    clipped_poly = cl.clip_shp(locs_buff, single_rect_poly_gdf)
    assert len(clipped_poly.geometry) == 3
    assert all(clipped_poly.geom_type == "Polygon")


def test_clip_multipoly(multi_gdf, single_rect_poly_gdf):
    """Test a multi poly object can be clipped properly.

    Also the bounds of the object should == the bounds of the clip object
    if they fully overlap (as they do in these fixtures). """
    clip = cl.clip_shp(multi_gdf, single_rect_poly_gdf)
    assert hasattr(clip, "geometry")
    assert np.array_equal(clip.total_bounds, single_rect_poly_gdf.total_bounds)
    # 2 features should be returned with an attribute column
    assert len(clip.attr) == 2


def test_clip_single_multipolygon(locs_buff, smaller_clip_rect_poly_gdf):
    """Test clipping a multi poly with another poly that

    no sliver shapes should be returned in this clip. """

    multi = locs_buff.dissolve(by="type").reset_index()
    clip = cl.clip_shp(multi, smaller_clip_rect_poly_gdf)

    assert hasattr(clip, "geometry") and clip.geom_type[0] == "Polygon"


def test_clip_multiline(multi_line, single_rect_poly_gdf):
    """Test that clipping a multiline feature with a poly returns expected output."""

    clip = cl.clip_shp(multi_line, single_rect_poly_gdf)
    assert hasattr(clip, "geometry") and clip.geom_type[0] == "MultiLineString"


def test_clip_multipoint(single_rect_poly_gdf, multi_point):
    """Clipping a multipoint feature with a polygon works as expected.

    should return a geodataframe with a single multi point feature"""

    clip = cl.clip_shp(multi_point, single_rect_poly_gdf)

    assert hasattr(clip, "geometry") and clip.geom_type[0] == "MultiPoint"
    assert hasattr(clip, "attr")
    # All points should intersect the clip geom
    assert all(clip.intersects(single_rect_poly_gdf.unary_union))


def test_clip_lines(linez_gdf, single_rect_poly_gdf):
    """Test what happens when you give the clip_extent a line GDF."""
    clip_line = cl.clip_shp(linez_gdf, single_rect_poly_gdf)
    assert len(clip_line.geometry) == 2
```

After refactoring that, I am also noticing that the fixture called `smaller_clip_rect_poly_gdf` could have a better name. If I understand correctly its actually **larger** than the other rectangle polygon gdf fixture, instead of smaller as the name might suggest.

Reply:
ok!! i'll be on mission fixture cleanup. i'll have a look at the `smaller_clip_rect_poly_gdf`. it's totally possible that it was opposite day when i created it. i wanted to create a poly that did not test the edge case of a clip of a mult-poly object that returns slivers (lines) and polygons. 

thanks @mbjoseph !!

Reply:
i just pushed a set of changes. i want to give this one more once over because i keep finding issues, duplications and points of clarification